### ISSUE TITLE
Admin color themes and customization?

### ISSUE THREAD

**Issue by scuerda** [wagtail/wagtail#2943]
I'm working on customizing the wagtail admin for a deployment and am finding it a bit challenging to catch all the locations where color schemes (background, font, hover, etc.) need to be overridden. 

The basic color variables are easy to redefine in the _variables.scss partial, but it's quite challenging to either rebuild core.css from scratch and/or to generate a custom-core.css file that doesn't contain a lot of duplicate styles. This is primarily due to the structure of the imports. I'd say that finding a way to avoid 

``` css
#wagtail {
    @import '../../../../../client/scss/style';
}
```

in core.scss would probably be the easiest first step in terms of being able to build a theming stylesheet.

Are there any roadmap items for addressing the need of users to do this type of customization and if so, are there any ways to help move this forward?

**Comment by gasman** [comment]
Hi @scuerda,
There's not currently any roadmap item for this - it isn't something we've needed. We'd be happy to consider a PR that implements it, though.

One major consideration here is that a standard installation of Wagtail (as installed through `pip install wagtail`) comes with the CSS files prebuilt, and will not include the Node-based toolchain for rebuilding them from SCSS. We presumably want to avoid overwriting files within the Wagtail package, so the neatest approach is probably to write the compiled CSS to an app `/static` directory within your Wagtail project - as long as that app appears above `wagtailadmin` etc in `INSTALLED_APPS`, those files should override the ones shipped with Wagtail. This in turn will involve some rewriting of the Node build scripts, and detailed documentation for setting it all up (since the steps for building the CSS go way beyond a "standard" Wagtail install).

**Comment by loleg** [comment]
I am not sure I see the need to customize everything, but it would be nice to have at least a way of specifying the admin header color to make it easier for users to tell production / staging / test servers apart.

**Comment by SalahAdDin** [comment]
:+1:

**Comment by JimBrouzoulis** [comment]
I'm looking to do some refactoring in the scss files to simplify the themeing process. This may be a step forward regarding this issue. 

I'm thinking of having something similar to many popular frameworks with a primary color, secondary etc. Wagtail's default theme would then use `$color-teal` and `$color-salmon`. 

However, I have a bit more difficulty understanding the usage of the greys: `$color-grey-1`-`$color-grey-5`. In total there are 6 defined greys and several more `lighten` and `darken` versions.

- Is there a semantic meaning to the different versions or is it mostly esthetics?
- Do we need that many definitions of grey, can't we make do with say three versions?

**Comment by gasman** [comment]
@JimBrouzoulis The developer who wrote the original CSS has left the project, so there probably won't be a definitive answer to that... I suspect it is mostly aesthetics, though.

**Comment by JimBrouzoulis** [comment]
Ok I see. Well I'm gonna give it a go and see what I come up with.

**Comment by Meteor0id** [comment]
Adjusting the editor interface colorscheme is definetly a missing feature. What I am hoping to see is a simple way to modify this just like we have for replacing the wagtail logo.
Using admin templates.

[https://docs.wagtail.io/en/v2.1.2/advanced_topics/customisation/admin_templates.html](https://docs.wagtail.io/en/v2.1.2/advanced_topics/customisation/admin_templates.html)

**Comment by Meteor0id** [comment]
I wouldn't worry with the gray just yet. For now let's focus on a way to let the user replace all instances of only the primairy and secondary color:

all occurences of:
#358c8b;
#43b1b0;

What I am not usre about is how we could give the user an easy way to define these. They are currently hardcoded all over the place. They could probably simply be replaced by a variable, but the user will need an easy way to overwrite that variable though the template system. And this is scss, not html, so I am not sure how to do that.

**Comment by Provoo** [comment]
There is still no correct way to change administrator colors?

**Comment by gasman** [comment]
@Provoo Please see the discussion above - there are technical complications that will need to be solved before this is possible.

**Comment by Provoo** [comment]
Yes, I understand but I'm surprise that before 2 years and half, this bug continuous, I really appreciate it and it seems to me an excellent wagtail tool, but you should warn that it does not have this feature in the docs, because we selected it for a corporate product, but they ask us for the administrator to have the colors and logo of the brand, and we have to do a lot of stuff to change only colors.

**Comment by thibaudcolas** [comment]
@Provoo I think this keeps on being an issue simply because it's easier for people who have this problem to make their own CSS / custom template customisations for their own project, instead of making PRs to introduce this as configuration within Wagtail.

I think it would be very valuable already if someone making those kinds of customisations documented the process, and published a blog post / tutorial about it.

---

As an alternative future solution, once Wagtail is able to remove browser support for IE11 we will be able to use [CSS variables / custom properties](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties) to define the UI's colors. This should make it very easy to override the admin UI's color theme in a consistent and exhaustive way. But that's only once IE11 stops being relevant, so probably a few years from now at least.

**Comment by csalmeida** [comment]
I am working on a project in which is required to make a lot of changes to how the admin panel looks. I have learned how to change the markup, but still wondering if anyone found their own effective way of styling the panel?

I have tried to add a `link` tag to the CSS block and using a hook without results so I would deeply appreciate any tips on this. 🙏

**Comment by easherma** [comment]
@thibaudcolas is anything stopping us from using css variables with a polyfill for IE11 support?

**Comment by thibaudcolas** [comment]
My preferred approach would be something like https://github.com/postcss/postcss-custom-properties, so we get to use CSS variables without making Wagtail even slower for IE11 users. With the caveat that theming won’t be supported in IE11, which seems ok to me.

For a JS polyfill – I’d like to see whether there is some consensus that it’s worthwhile to compromise performance for this.

----

Either way my assumption from https://github.com/wagtail/wagtail/issues/2943#issuecomment-474953477 is that theming isn’t a feature that anyone is interested in enough to invest into that refactoring – I’d assume we’ll get it "for free" once IE11 is no longer relevant and other UI improvements (/ refactoring #3804) justify the effort of moving to CSS variables.

**Comment by thibaudcolas** [comment]
For people interested in this, we now have an issue to track / discuss removing support for IE11: #6170. 

This alone won’t make Wagtail any more theme-able, but again it’ll allow us to start using CSS variables in Wagtail’s stylesheets without much effort, which will make theming _much_ easier for colours (and fonts, and other design tokens, if relevant).

I can see this happening like this:

- Refactoring some (or all) of our existing Sass variables colors to CSS variables (custom properties)
- Documenting on https://docs.wagtail.io/en/stable/advanced_topics/customisation/admin_templates.html which color variables are available
- Additionally having a few words about adding a custom stylesheet that implements those overrides.


If anyone wants to see this happen sooner, I’d also welcome a proposal (/ WIP PR) to explore what this looks like with https://github.com/postcss/postcss-custom-properties for IE11 support. 

---

An added benefit of this approach is that it should also make it very easy for us to add "dark mode" support to Wagtail… no one has asked for it yet, but personally I think it’d be quite cool 😬 .

**Comment by JNaftali** [comment]
> If anyone wants to see this happen sooner, I’d also welcome a proposal (/ WIP PR) to explore what this looks like with https://github.com/postcss/postcss-custom-properties for IE11 support.

I took a swing at this! https://github.com/wagtail/wagtail/pull/6349

**Comment by thibaudcolas** [comment]
Thank you to @JNaftali for making this happen via #6409 🙏 We can now customize 5 of Wagtail’s colors via very straightforward CSS custom properties definitions. See https://docs.wagtail.io/en/latest/advanced_topics/customisation/admin_templates.html for the full documentation.

This will be released in [Wagtail 2.12](https://github.com/wagtail/wagtail/wiki/Release-schedule).

I will close this now – if anyone wants to see more things being customizable in the admin I think this will warrant a new issue. As much as possible, make sure to suggest specific customizations, provide the clear use case you have for them, with screenshots if applicable.

**Comment by thibaudcolas** [comment]
Wagtail 2.12rc1 has now been released with this feature, if anyone wants to give it a go: https://pypi.org/project/wagtail/2.12rc1/. Release notes: https://docs.wagtail.io/en/latest/releases/2.12.html#admin-color-themes
